src/async-wrap.cc

```diff
@@ -407,6 +407,40 @@ static void DisablePromiseHook(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+class AsyncWrap::DestroyParam {
+ public:
+  double asyncId;
+  v8::Persistent<Function> callback;
+};
+void AsyncWrap::WeakCallback(
+  const v8::WeakCallbackInfo<AsyncWrap::DestroyParam> &info
+) {
+  AsyncWrap::DestroyParam* p = info.GetParameter();
+  Local<Function> fn = PersistentToLocal(info.GetIsolate(), p->callback);
+  Local<v8::Value> asyncId = v8::Number::New(info.GetIsolate(), p->asyncId);
+  delete p;
+
+  Local<v8::Primitive> undef = Undefined(info.GetIsolate());
+  fn->Call(
```

Comment:
I'm not entirely sure how this `SetWeek` works but we have had issues before, where calling into JavaScript when the GC is running caused a process crash. This is why we have `PushBackDestroyAsyncId`.

Reply:
Did you delete your other comment? I think it'd be possible to do this rather easily, if we save a Map<AsyncId, destroyManual> in JS, but I can't quite figure out how `MakeCallback` works for the other thing you mentioned.

Reply:
It got collapsed https://github.com/nodejs/node/pull/16998#discussion_r150668918

Reply:
If make it such that it calls `emitNativeDestroy` directly (some suggestions in #16998 (comment)) then it should be straightforward to use `PushBackDestroyAsyncId`.

`PushBackDestroyAsyncId` basically just calls `emitNativeDestroy` when it is safe.